lib/publish.js

```diff
@@ -1,30 +1,97 @@
+const {createReadStream} = require('fs');
+const {resolve} = require('path');
+const {stat} = require('fs-extra');
+const FormData = require('form-data');
 const urlJoin = require('url-join');
 const got = require('got');
 const debug = require('debug')('semantic-release:gitlab');
 const resolveConfig = require('./resolve-config');
 const getRepoId = require('./get-repo-id');
+const getAssets = require('./get-assets');
 
 module.exports = async (pluginConfig, context) => {
   const {
+    cwd,
     options: {repositoryUrl},
     nextRelease: {gitTag, gitHead, notes},
     logger,
   } = context;
-  const {gitlabToken, gitlabUrl, gitlabApiPathPrefix} = resolveConfig(pluginConfig, context);
+  const {gitlabToken, gitlabUrl, gitlabApiPathPrefix, assets} = resolveConfig(pluginConfig, context);
   const repoId = getRepoId(gitlabUrl, repositoryUrl);
   const encodedRepoId = encodeURIComponent(repoId);
-  const apiUrl = urlJoin(gitlabUrl, gitlabApiPathPrefix);
   const encodedGitTag = encodeURIComponent(gitTag);
+  const apiBase = urlJoin(gitlabUrl, gitlabApiPathPrefix);
+  const apiProject = urlJoin(apiBase, `/projects/${encodedRepoId}`);
+  const apiOptions = {
+    json: true,
+    headers: {
+      'PRIVATE-TOKEN': gitlabToken,
+    },
+  };
+  let finalNotes = notes;
 
   debug('repoId: %o', repoId);
   debug('release name: %o', gitTag);
   debug('release ref: %o', gitHead);
 
+  if (assets && assets.length > 0) {
+    const globbedAssets = await getAssets(context, assets);
+    const assetsMd = [];
+    debug('globed assets: %o', globbedAssets);
+
+    await Promise.all(
+      globbedAssets.map(async asset => {
+        const filePath = asset.path;
+        let fileStat;
+        let file;
+
+        try {
+          file = resolve(cwd, filePath);
+          fileStat = await stat(file);
+        } catch (error) {
+          logger.error('The asset %s cannot be read, and will be ignored.', filePath);
+          return;
+        }
+        if (!fileStat || !fileStat.isFile()) {
+          logger.error('The asset %s is not a file, and will be ignored.', filePath);
+          return;
+        }
+
+        const fileName = asset.name;
+        debug('file path: %o', filePath);
+        debug('file name: %o', fileName);
+
+        // Uploaded assets to the project
+        let body;
+        try {
+          const form = new FormData();
+          form.append('file', createReadStream(file));
+
+          ({body} = await got.post(urlJoin(apiProject, '/uploads'), {
+            ...apiOptions,
+            json: false,
+            body: form,
+          }));
+          body = JSON.parse(body);
+        } catch (error) {
```

Comment:
in the case of 400?

Reply:
No. But if you have a 400 it means you a bug in the code. You don't want to skip assets because of a bug here, you want to make sure your request is correct.

Reply:
removed